File: C++/VI to IX - Programming Exercise (Sorting Benchmarks).cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>
#include <ctime>

using namespace std;

int bubbleSort(vector<int> arr) {
    int swaps = 0;
    for (int i = 0; i < arr.size(); i++) {
        for (int j = 0; j < arr.size() - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                swaps++;
            }
        }
    }
    return swaps;
}

int selectionSort(vector<int> arr) {
    int swaps = 0;
    for (int i = 0; i < arr.size(); i++) {
        int min = i;
        for (int j = i + 1; j < arr.size(); j++) {
            if (arr[j] < arr[min]) {
                min = j;
            }
        }
        if (min != i) {
            int temp = arr[i];
            arr[i] = arr[min];
            arr[min] = temp;
            swaps++;
        }
    }
    return swaps;
}

int insertionSort(vector<int> arr) {
    int swaps = 0;
    for (int i = 1; i < arr.size(); i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && key < arr[j]) {
            arr[j + 1] = arr[j];
            j--;
            swaps++;
        }
        arr[j + 1] = key;
    }
    return swaps;
}

int binary(vector<int> array, int value) {
    int count = 0;
    int low = 0;
    int high = array.size() - 1;
    while (low <= high) {
        count += 1;
        int mid = (low + high) / 2;
        if (array[mid] == value) {
            return count;
        } else if (array[mid] < value) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return -1;
}

int main() {
    string name;
    vector<int> array;
    srand(time(NULL));
    for (int num = 0; num < 20; num++) {
        array.push_back(rand() % 100 + 1);
    }
    sort(array.begin(), array.end());
    cout << "Full Name: ";
    getline(cin, name);
    cout << "Benchmark Array: ";
    for (int i = 0; i < array.size(); i++) {
        cout << array[i] << " ";
    }
    cout << endl;
    vector<int> arr1;
    vector<int> arr2;
    vector<int> arr3;
    for (int i = 0; i < 20; i++) {
        arr1.push_back(rand() % 100);
        arr2.push_back(rand() % 100);
        arr3.push_back(rand() % 100);
    }
    cout << "Bubble Sort Swap Count: " << bubbleSort(arr1) << endl;
    cout << "Selection Sort Swap Count: " << selectionSort(arr2) << endl;
    cout << "Insertion Sort Swap Count: " << insertionSort(arr3) << endl;
    return 0;
}

File: Python/VI to IX - Programming Exercise (Sorting Benchmarks).py
import random

def bubbleSort(arr):
    n = len(arr)
    count = 0
    for i in range(n):
        for j in range(0, n-i-1):
            if arr[j] > arr[j+1] :
                arr[j], arr[j+1] = arr[j+1], arr[j]
                count += 1
    return count

def selectionSort(arr):
    count = 0
    for i in range(len(arr)):
        min_idx = i
        for j in range(i+1, len(arr)):
            if arr[min_idx] > arr[j]:
                min_idx = j
                count += 1
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return count

def insertionSort(arr):
    count = 0
    for i in range(1, len(arr)):
        key = arr[i]
        j = i-1
        while j >=0 and key < arr[j] :
            arr[j+1] = arr[j]
            j -= 1
            count += 1
        arr[j+1] = key
    return count

def main():
    array = []
    for i in range(20):
        array.append(random.randint(1, 100))
    array.sort()
    str_name = str(input("Full Name: "))
    print("Benchmark Array:",array)
    
    arr1 = [random.randint(0, 100) for i in range(20)]
    arr2 = [random.randint(0, 100) for i in range(20)]
    arr3 = [random.randint(0, 100) for i in range(20)]
    
    print("Bubble Sort Swap Count: ", bubbleSort(arr1))
    print("Selection Sort Swap Count: ", selectionSort(arr2))
    print("Insertion Sort Swap Count: ", insertionSort(arr3))

if __name__ == "__main__":
    main()
